### ISSUE TITLE
exchangelib.errors.ErrorFolderNotFound raised if well-known folder is missing from mailbox

### ISSUE THREAD

**Issue by bn-crandom** [ecederstrand/exchangelib#1155]
**Describe the bug**
We have some customers with mailboxes that are somehow missing some well-known folders, e.g. Conversation History.  Our application does not have permission to create folders.  If we try to enumerate the folders in one of these mailboxes, we get a exchangelib.errors.ErrorFolderNotFound exception resulting from an upstream Microsoft.Exchange.Services.Core.Types.ErrorFolderNotFoundException error.

exchangelib will deal with ErrorFolderNotFound exceptions if they occur for individual folders, but not if that error is returned for the  GetFolders request itself.

**To Reproduce**
This isn't easily reproducible.  I deleted some well-known folders from one of our dev mailboxes and didn't get this error. 

**Expected behavior**
exchangelib should be able to enumerate folders for these mailboxes.  Our workaround is to force resolution of all well-known folders before enumerating folder in the mailbox.  If we catch a ErrorFolderNotFound exception, we resolve each well-known folder individually.  Each well-known folder that cannot be resolved is removed from the mailbox's list of well-known folders.  We then retry resolving the list of well-known folders again.  This has been successful.
```
 def resolve_distinguished_folders(self):
        def init_cache(root):
            # This is an "internal" method, but there's no explicit way to invoke distinguished folder resolution
            root._folders_map

        try:
            # This will resolve all distinguished folders in one EWS operation
            logging.info("Resolving distinguished folders...")
            init_cache(self.account.root)
            logging.info("Folders resolved")
        except (ErrorFolderNotFound, ErrorInvalidOperation):
            # We failed due to a missing distinguished folder; try to resolve each distinguished folder individually
            # Add the distinguished folders that do exist to a new list
            existing_wellknown_folders = []
            logging.error("Caught exception while resolving distinguished folders; identifying missing folders...")
            for cls in self.account.root.WELLKNOWN_FOLDERS:
                try:
                    logging.info(f"Resolving distinguished folder: {cls}")
                    cls.get_distinguished(self.account.root)
                    existing_wellknown_folders.append(cls)
                except Exception as e:
                    logging.error(f"Failed to resolve distinguished folder '{cls}': {repr(e)}")

            # Replace this account's root folder's list of distinguished folders with the new filtered list
            self.account.root.WELLKNOWN_FOLDERS = existing_wellknown_folders

            # Reset the root folder's cache
            self.account.root.clear_cache()

            # Retry the resolution of the distinguished folders
            logging.info("Collected list of existing distinguished folders; retrying resolution...")
            init_cache(self.account.root)
```

**Log output**
```
2022-09-21 00:07:38,432: INFO    1/140656002523792 processor.py:98: Processing mailbox: <omitted>
2022-09-21 00:07:39,283: ERROR   1/140656002523792 cli.py:329: Caught exception while processing mailbox:
Traceback (most recent call last):
  File "/pyxint/pyxint/pyxint/cli.py", line 291, in enumerate_mailboxes
    self.process_mailbox(mailbox, configuration, job_config.args.access_type)
  File "/pyxint/pyxint/pyxint/cli.py", line 272, in process_mailbox
    self.mailbox_processor.process_mailbox(account)
  File "/pyxint/pyxint/pyxint/processor.py", line 101, in process_mailbox
    self.process_mailbox_contents(account)
  File "/pyxint/pyxint/pyxint/processor.py", line 257, in process_mailbox_contents
    self.process_all_contents(account)
  File "/pyxint/pyxint/pyxint/processor.py", line 234, in process_all_contents
    self.process_folder(account.msg_folder_root)
  File "/pyxint/pyxint/pyxint/processor.py", line 314, in process_folder
    if self.skip_folder(folder):
  File "/pyxint/pyxint/pyxint/pim_importer.py", line 83, in skip_folder
    logging.info(f"Folder container class does not match filter: {format_folder_path(self.mailbox, folder, self.public_folders)} is class {folder.folder_class}; skipping")
  File "/pyxint/pyxint/pyxint/util.py", line 68, in format_folder_path
    if folder.absolute.startswith(ABSOLUTE_FOLDER_PATH_PREFIX):
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/folders/base.py", line 103, in absolute
    return ''.join('/%s' % p.name for p in self.parts)
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/folders/base.py", line 95, in parts
    f = self.parent
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/folders/base.py", line 633, in parent
    return self.root.get_folder(self.parent_folder_id.id)
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/folders/roots.py", line 71, in get_folder
    return self._folders_map.get(folder_id, None)
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/folders/roots.py", line 159, in _folders_map
    for f in FolderCollection(account=self.account, folders=distinguished_folders).resolve():
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/folders/collections.py", line 276, in resolve
    additional_fields=additional_fields
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/folders/collections.py", line 342, in get_folders
    shape=ID_ONLY,
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/services/get_folder.py", line 35, in call
    shape=shape,
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/services/common.py", line 550, in _pool_requests
    for elem in self._get_elements(payload=payload_func(chunk, **kwargs)):
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/services/common.py", line 118, in _get_elements
    for i in self._response_generator(payload=payload):
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/services/common.py", line 109, in _response_generator
    response = self._get_response_xml(payload=payload)
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/services/common.py", line 207, in _get_response_xml
    return self._get_soap_messages(body=body, **parse_opts)
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/services/common.py", line 293, in _get_soap_messages
    cls._raise_soap_errors(fault=fault)  # Will throw SOAPError or custom EWS error
  File "/root/.pyenv/versions/3.7.8/lib/python3.7/site-packages/exchangelib/services/common.py", line 329, in _raise_soap_errors
    raise vars(errors)[code](msg)
exchangelib.errors.ErrorFolderNotFound: Microsoft.Exchange.Services.Core.Types.ErrorFolderNotFoundException: The specified folder could not be found in the store. ---> Microsoft.Exchange.Data.StoreObjects.CannotCreateDefaultFolderException: The defaultFolder cannot be created by XSO (it does not have CanCreate behavior). defaultFolder = Conflicts, IsDelegatedLogon = False, IsOwnerLogon = True, Is defaultFolderData = HasInitialized=True,
IdInitialized=True,
InitializationException=,
FolderId=,
.
   at Microsoft.Exchange.Data.StoreObjects.DefaultFolders.DefaultFolderDataExtensions.ThrowIfCannotCreateFolder(DefaultFolderData defaultFolderData, DefaultFolderInfo defaultFolderInfo, ISessionProperties sessionProperties)
   at Microsoft.Exchange.Data.StoreObjects.DefaultFolders.DefaultFolderContextExtensions.Create(IDefaultFolderContext defaultFolderContext, DefaultFolderData defaultFolderData, DefaultFolderInfo defaultFolderInfo)
   at Microsoft.Exchange.Data.StoreObjects.DefaultFolders.DefaultFolderContextExtensions.SetDefaultFolderData(IDefaultFolderContext defaultFolderContext, DefaultFolderType defaultFolderType, GetDefaultFolderDataDelegate getDefaultFolderData)
   at Microsoft.Exchange.Data.StoreObjects.DefaultFolders.DefaultFolderContextExtensions.GetOrCreateDefaultFolder(IDefaultFolderContext defaultFolderContext, DefaultFolderType defaultFolderType)
   at Microsoft.Exchange.Data.StoreObjects.DefaultFolders.DefaultFolderManager.GetOrCreate(DefaultFolderType type)
   at Microsoft.Exchange.Data.Storage.MailboxSession.GetOrCreateDefaultFolder(DefaultFolderType defaultFolderType)
   at Microsoft.Exchange.Services.Core.Types.SessionAndAuthZ.GetRefreshedDefaultFolderId(DefaultFolderType defaultFolderType)
   --- End of inner exception stack trace ---
   at Microsoft.Exchange.Services.Core.Types.SessionAndAuthZ.GetRefreshedDefaultFolderId(DefaultFolderType defaultFolderType)
   at Microsoft.Exchange.Services.Core.Types.ExtensionMethods.GetRefreshedDefaultFolderId(IStoreSession session, DefaultFolderType defaultFolderType, IMinimalCallContext minimalCallContext)
   at Microsoft.Exchange.Services.Core.Types.IdConverter.ConvertDefaultFolderType(IMinimalCallContext callContext, DefaultFolderType defaultFolderType, IStoreSession sessionToUse)
   at Microsoft.Exchange.Services.Core.Types.IdConverter.ConvertDefaultFolderType(IMinimalCallContext callContext, DefaultFolderType defaultFolderType, String emailAddress, Boolean archiveSession)
   at Microsoft.Exchange.Services.Core.Types.IdConverter.ConvertDistinguishedFolderId(IMinimalCallContext callContext, String displayName, String changeKey, String emailAddress, ConvertOption convertOption)
   at Microsoft.Exchange.Services.Common.FolderIdConverter.ConvertFolderIdToIdAndSessionAndGetCorrespondingFolder(BaseFolderId folderId, ConvertOption convertOption, IFolder& folder)
   at Microsoft.Exchange.Services.Common.FolderIdConverter.ConvertFolderIdToIdAndSession(BaseFolderId folderId, ConvertOption convertOption)
   at Microsoft.Exchange.Services.Core.GetFolder.Execute()
   at Microsoft.Exchange.Services.Core.ExceptionHandler`1.Execute(IMinimalCallContext callContext, ISimpleLogger logger, Func`2 createServiceResult, Int32 index, ExecutionOption executionOption)
```

**Additional context**
Python 3.7.8
exchangelib 4.7.3

**Comment by ecederstrand** [comment]
Thanks for the report!

If you look at https://github.com/ecederstrand/exchangelib/blob/master/exchangelib/folders/roots.py#L162, exchange lib should catch `ErrorFolderNotFound` because it's a member of `MISSING_FOLDER_ERRORS`, and because `ErrorFolderNotFound` is a member of ` ERRORS_TO_CATCH_IN_RESPONSE` (https://github.com/ecederstrand/exchangelib/blob/master/exchangelib/services/get_folder.py#L11) which means that the error is passed to the caller, not raised.

exchange lib is supposed to handle your situation gracefully. So I wonder why this happens anyway in your case? Are you on the latest version of exchangelib? Can you attach a debugger and find out why it happens?

**Comment by bn-crandom** [comment]
@ecederstrand, I traced through the code and saw that these exceptions are supposed to be intercepted, which confused me.  I found that exchangelib will deal with `ErrorFolderNotFound` exceptions if they occur for individual folders, but not if that error is returned for the `GetFolders` request itself.  I.e. if `resolve` in the line you linked returned a mix of exceptions and folders, it'll handle the exceptions, but it doesn't handle any of the `MISSING_FOLDER_ERRORS` if `resolve` _itself_ throws.

We were on exchangelib 4.7.3 when we encountered this, but I don't see any changes that would fix this.  We can try with the newer version, but I expect that this is still a problem.
